### PR TITLE
Simcomp for force/torque calculation 

### PR DESCRIPTION
**Pull request by MartinKarp**
SImcomp to calculate the force and torque around a zone. Nothing too fancy, but it does the trick.

Also adds device_vcross as well as the possibility to copy data according to a mask to a reduced length array.

### CODE REVIEW

**Review by njansson — APPROVED** [review]

File: src/qoi/drag_torque.f90 [hunk 1]
```diff
@@ -59,20 +59,25 @@
...
 
 module drag_torque
-  use field, only : field_t
-  use coefs, only : coef_t
-  use mesh, only : mesh_t
-  use facet_zone, only : facet_zone_t
+  use field, only: field_t
+  use coefs, only: coef_t
+  use mesh
+  use facet_zone
```
Comment (MartinKarp) [inline src/qoi/drag_torque.f90:65]: Will fix this.

File: src/math/bcknd/device/cuda/math_kernel.h [hunk 2]
```diff
@@ -50,6 +50,25 @@ __global__ void cmult_kernel(T * __restrict__ a,
...
 }
 
+
+/**
+ * Device kernel for masked red copy
+ */
+template< typename T >
+__global__ void masked_red_copy_kernel(T * __restrict__ a,
+                                   T * __restrict__ b,
+                                   int * __restrict__ mask,
```
Comment (timfelle) [inline src/math/bcknd/device/cuda/math_kernel.h:60]:
```suggestion
                                   int * __restrict__ mask,
```

File: src/qoi/drag_torque.f90 [hunk 3]
```diff
@@ -148,7 +153,7 @@ subroutine drag_torque_zone(dgtq, tstep, zone, center, s11, s22, s33, s12, s13,
          torqvx = torqvx + dgtq(1,4)  ! viscous
          torqvy = torqvy + dgtq(2,4)
          torqvz = torqvz + dgtq(3,4)
-      end do
+      enddo
```
Comment (timfelle) [inline src/qoi/drag_torque.f90:156]: It should have a space here :sweat_smile:
```suggestion
      end do
```

File: src/simulation_components/force_torque.f90 [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    real(kind=rp) :: scale
+    
+    ! Add fields keyword to the json so that the field_writer picks it up.
+    ! Will also add fields to the registry.
+
+    call this%init_base(json, case)
+
+    call json_get(json, 'zone_id',zone_id)
+    call json_get_or_default(json, 'zone_name',zone_name,' ')
+    call json_get_or_default(json, 'scale',scale,1.0_rp)
```
Comment (timfelle) [inline src/simulation_components/force_torque.f90:119]:
```suggestion
    call json_get(json, 'zone_id', zone_id)
    call json_get_or_default(json, 'zone_name', zone_name, ' ')
    call json_get_or_default(json, 'scale', scale, 1.0_rp)
```

**Review by timfelle — APPROVED** [review]

File: src/simulation_components/force_torque.f90 [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+     !> X velocity component.
+     type(field_t), pointer :: u
+     !> Y velocity component.
+     type(field_t), pointer :: v
+     !> Z velocity component.
+     type(field_t), pointer :: w
+     !> Pressure.
+     type(field_t), pointer :: p
+
+     type(field_t) :: s11, s22, s33, s12, s13, s23
```
Comment (timofeymukha) [inline src/simulation_components/force_torque.f90:73]: Should probably use the scratch registry for these.

File: src/simulation_components/force_torque.f90 [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+  use math, only : masked_red_copy, cadd, glsum, vcross
+  use device_math, only : device_masked_red_copy, device_cadd, &
+                          device_glsum, device_vcross
+  use device
+  
+  implicit none
+  private
+
+  !> A simulation component that computes the force_torque field.
+  !! Added to the field registry as `omega_x`, `omega_y``, and `omega_z`.
```
Comment (timofeymukha) [inline src/simulation_components/force_torque.f90:62]: Incorrect docstring regarding added fields

File: src/simulation_components/force_torque.f90 [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    call json_get_or_default(json, 'zone_name', zone_name, ' ')
+    call json_get_or_default(json, 'scale', scale, 1.0_rp)
+    call json_get(json, 'center', center)
+    call force_torque_init_from_attributes(this, zone_id, zone_name, &
+                                           center, scale, case%fluid%c_xh)
+  end subroutine force_torque_init_from_json
+
+  !> Actual constructor.
+  subroutine force_torque_init_from_attributes(this, zone_id, zone_name, &
+                                               center, scale, coef)
```
Comment (timofeymukha) [inline src/simulation_components/force_torque.f90:127]: params

**Review by timofeymukha — COMMENTED** [review]
Late to the party, but a few comments.

Also, the flint scores are a bit criminal :-).

	- src/qoi/drag_torque.f90: 6.56
	- src/simulation_components/force_torque.f90: 7.35

**Pull request merged**